server/migrations/20190410170255-cache-root-workspace.js

```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+
+module.exports = {
+  up: async function(queryInterface, Sequelize) {
+    async function findRootWorkspaceId(workspaceId) {
+      const result = await queryInterface.sequelize.query(
+        `SELECT "parentId" FROM "Workspaces" WHERE "id" = :id`,
+        {
+          replacements: {
+            id: workspaceId,
+          },
+          type: Sequelize.QueryTypes.SELECT,
+        }
+      );
+
+      if (result.length === 0) {
+        return workspaceId;
+      }
+
+      const parentId = result[0].parentId;
+      if (parentId === null) {
+        return workspaceId;
+      } else {
+        return await findRootWorkspaceId(parentId);
+      }
+    }
+
+    await queryInterface.addColumn("Workspaces", "rootWorkspaceId", {
+      type: Sequelize.UUID,
+      references: {
+        model: "Workspaces",
+        key: "id"
+      }
+    });
+
+    const workspaces = await queryInterface.sequelize.query(
+      `SELECT "id" FROM "Workspaces"`,
+      {
+        type: Sequelize.QueryTypes.SELECT,
+      }
+    );
+
+    await Promise.all(workspaces.map(async ({ id }) => {
+      const rootId = await findRootWorkspaceId(id);
+      await queryInterface.sequelize.query(
+        `UPDATE "Workspaces" SET "rootWorkspaceId" = :rootId`,
```

Comment:
This needs a `WHERE` clause. Right now every time it's re-updating every workspace.

Reply:
😬 good catch